Common/cpp/SharedItems/Shareables.h

```diff
@@ -0,0 +1,270 @@
+#pragma once
+
+#include <jsi/jsi.h>
+#include <memory>
+
+#include "RuntimeManager.h"
+#include "Scheduler.h"
+
+using namespace facebook;
+
+namespace reanimated {
+
+class CoreFunction;
+
+class JSRuntimeHelper {
+ public:
+  jsi::Runtime *uiRuntime; // UI runtime created by Reanimated
+  jsi::Runtime *rnRuntime; // React-Native's main JS runtime
+  std::shared_ptr<Scheduler> scheduler;
+
+  JSRuntimeHelper(
+      jsi::Runtime *_rnRuntime,
+      jsi::Runtime *_uiRuntime,
+      const std::shared_ptr<Scheduler> &_scheduler)
+      : rnRuntime(_rnRuntime), uiRuntime(_uiRuntime), scheduler(_scheduler) {}
+
+  std::shared_ptr<CoreFunction> valueUnpacker;
+
+  inline bool isUIRuntime(const jsi::Runtime &rt) const {
+    return &rt == uiRuntime;
+  }
+
+  inline bool isRNRuntime(const jsi::Runtime &rt) const {
+    return &rt == rnRuntime;
+  }
+
+  void scheduleOnUI(std::function<void()> job) {
+    scheduler->scheduleOnUI(job);
+  }
+
+  void scheduleOnJS(std::function<void()> job) {
+    scheduler->scheduleOnJS(job);
+  }
+};
+
+// Core functions are not allowed to capture outside variables, otherwise they'd
+// try to access _closure variable which is something we want to avoid for
+// simplicity reasons.
+class CoreFunction {
+ private:
+  std::shared_ptr<jsi::Function> rnFunction;
+  std::shared_ptr<jsi::Function> uiFunction;
+  std::string functionBody;
+  std::string location;
+  JSRuntimeHelper
+      *runtimeHelper; // runtime helper holds core function references, so we
+                      // use normal pointer here to avoid ref cycles.
+ public:
+  CoreFunction(
+      JSRuntimeHelper *_runtimeHelper,
+      const jsi::Value &workletObject);
+  jsi::Value call(jsi::Runtime &rt, const jsi::Value &arg0);
+};
+
+class Shareable {
+ protected:
+  virtual jsi::Value toJSValue(jsi::Runtime &rt) = 0;
+
+ public:
+  virtual ~Shareable();
+
+  enum ValueType {
+    UndefinedType,
+    NullType,
+    BooleanType,
+    NumberType,
+    // SymbolType, TODO
+    // BigIntType, TODO
+    StringType,
+    ObjectType,
+    ArrayType,
+    WorkletType,
+    RemoteFunctionType,
+    HandleType,
+  };
+
+  Shareable(ValueType valueType_) : valueType(valueType_){};
```

Comment:
The convention (at least in React Native C++ codebase) is that we mark instance fields with trailing underscore (e.g. `foo_`) and don't mark constructor args with underscore, so effectively the initializer list items looks like this: `foo_(foo)`.

We are incrementally adopting this convention in Reanimated sources as well, especially since the Fabric migration.

Can we apply this convention in this PR? 

References:
* https://github.com/facebook/react-native/blob/089684ee5604a6a2a15a4fc03fe63c73a5de9fd0/ReactCommon/react/renderer/mounting/MountingTransaction.h#L81-L84
* https://github.com/facebook/react-native/blob/089684ee5604a6a2a15a4fc03fe63c73a5de9fd0/ReactCommon/react/renderer/mounting/MountingTransaction.cpp#L19-L22